src/zon.zig

```diff
@@ -0,0 +1,661 @@
+    },
+    none,
+
+    fn init(ty: ?Type, sema: *Sema) !@This() {
+        const t = ty orelse return .none;
+        const ip = &sema.pt.zcu.intern_pool;
+        switch (t.zigTypeTagOrPoison(sema.pt.zcu) catch return .none) {
+            .Struct => {
+                try t.resolveFully(sema.pt);
+                const loaded_struct_type = ip.loadStructType(t.toIntern());
+                return .{ .st = .{
+                    .ty = t,
+                    .loaded = loaded_struct_type,
+                } };
+            },
+            .Union => {
+                try t.resolveFully(sema.pt);
+                const loaded_union_type = ip.loadUnionType(t.toIntern());
+                const loaded_tag_type = loaded_union_type.loadTagType(ip);
+                return .{ .un = .{
+                    .ty = t,
+                    .loaded = loaded_tag_type,
+                } };
+            },
+            else => return .none,
+        }
+    }
+
+    fn get(self: *const @This(), name: NullTerminatedString, zcu: *Zcu) ?Type {
+        const ip = &zcu.intern_pool;
+        const self_ty, const index = switch (self.*) {
+            .st => |st| .{ st.ty, st.loaded.nameIndex(ip, name) orelse return null },
+            .un => |un| .{ un.ty, un.loaded.nameIndex(ip, name) orelse return null },
+            .none => return null,
+        };
+        return self_ty.structFieldType(index, zcu);
+    }
+};
+
+fn expr(self: LowerZon, node: Ast.Node.Index, res_ty: ?Type) !InternPool.Index {
+    const gpa = self.sema.gpa;
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const data = self.file.tree.nodes.items(.data);
+    const tags = self.file.tree.nodes.items(.tag);
+    const main_tokens = self.file.tree.nodes.items(.main_token);
+
+    // If the result type is slice, and our AST Node is not a slice, recurse and then take the
+    // address of the result so attempt to coerce it into a slice.
+    if (res_ty) |rt| {
+        const result_is_slice = rt.isSlice(self.sema.pt.zcu);
+        const ast_is_pointer = switch (tags[node]) {
+            .string_literal, .multiline_string_literal => true,
+            else => false,
+        };
+        if (result_is_slice and !ast_is_pointer) {
+            const val = try self.expr(node, rt.childType(self.sema.pt.zcu));
+            const val_type = ip.typeOf(val);
+            const ptr_type = try self.sema.pt.ptrTypeSema(.{
+                .child = val_type,
+                .flags = .{
+                    .alignment = .none,
+                    .is_const = true,
+                    .address_space = .generic,
+                },
+            });
+            return ip.get(gpa, self.sema.pt.tid, .{ .ptr = .{
+                .ty = ptr_type.toIntern(),
+                .base_addr = .{ .anon_decl = .{
+                    .orig_ty = ptr_type.toIntern(),
+                    .val = val,
+                } },
+                .byte_offset = 0,
+            } });
+        }
+    }
```

Comment:
This is the logic that's problematic per my prior comment about coercing during lowering vs after lowering. Instead, `res_ty` should be considered by all possible branches below, performing appropriate coercions / refs.